docs/extensions/model-merger.md

```diff
@@ -0,0 +1,84 @@
+# Model Merger
+
+The model merger is used to automatically merge two models. Merge is not meant in the same way as in VCS systems like git. At the moment its main use case is incremental imports, although it is not limited to it.
+## Model Merge Specification 
+
+The model merge needs a customizable specification of how to merge instances of a given concept. The DSL should support the user to write correct specifications for a set of concepts and allow to reuse specifications for instance when a concept is extended or interfaces are implemented. Putting this into a DSL rather than writing this in plain java allows us to evolve the runtime and the language independently. Inspirations can be taken from the new VCS aspect in MPS. While a much simpler use case it might still be useful input.
+
+### Concrete Syntax Example
+```
+Spec
+---
+<Concept> -> Id : (node) = node.name
+                         ---- Properties ----
+                         name -> left | right | manual (left, right, node) -> <function returning the new property value>
+                        prop 2 -> 
+                        ---- Children ----
+                        childA [0..1] -> left | right | drop | auto | manual (left, right, node) -> <function returning the node used for the child>
+                        child [1] -> left | right | auto | manual 
+                        child [0..n] -> 
+                                               new on the right : add | drop
+                                               only existing on the the left: keep | drop 
+                                               element:  left | right | drop | auto | manual 
+```
+
+We explicitly don't use terms like ours/theirs, new/old or existing/incoming because we don't want to use terms associated with VCS use cases.
+
+### Operations 
+
+For all types of merges specification entries three operations are always available: *left*, *right* and *manual*. The semantics are largely the same no matter which part of a concept they are applied to.
+
+*left*: Changes on the left side of the merge dominate over changes on the right side. The result always contains the same state as passed as the left side,
+
+*right*: Changes on the right side of the merge dominate over changes on the left side. The result always contains the same state as passed as the right side.
+
+*manual*: The manual operation allows fine grained control about the the merge result. It allows to specify the merge result as a function. The function has three arguments: _left_, _right_ and _node_. The type of _left_ and _right_ depends on the part of the concept that is merged. For properties its the value of the property on the left and the right side of the merge. For a child role it's the child node on the left and right side of the merge. The _node_ argument always contained the complete node that is merged, for properties the node owning the property and for a child role the parent. 
```

Comment:
I‘m not sure if the single `node` parameter is enough here. If we only pass one node we should clearly stat in which state the node is as this point. Maybe we need to pass three nodes (_left_, _right_, _merged_) where the _left_/_right_ are the parent in the state of left or rite side of the merge. The _merged_ is the node in the current state after the merge has happened, that means it might be partially merged. One thing that came to my mind is that we should document in which order the individual parts of a merge specification are executed.

Reply:
> . One thing that came to my mind is that we should document in which order the individual parts of a merge specification are executed.

At first sight, I think:
1. We merge all properties. The order that we merge is to be defined, but I can think of at least two options:

- We merge per property. This means we would first i.e., merge all "name" properties from INamedConcept. The order that the different "properties" would be merged would have to be defined.
- We merge per concept. This means we iterate over the concepts to be merged and for each, we merge its properties. In this case, the merge order of the concepts is to be defined. I guess the order of the properties as well.

2. We merge children. Again more or less the same questions from above are applicable.

3. We merge references. I think it makes sense to always run this as last. 

As a more general rule, I think we can either:
1. Always run "automatic" migrations first. This means that for manual migrations, the input node you mentioned would be the merged node at the state where the automatic migrations where already performed.

2. Always run "manual" migrations first, meaning that the user always relies on the original state of the nodes to be merged (Which I actually think is a more natural approach).